src/base/static/components/input-form/index.js

```diff
@@ -0,0 +1,551 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { fromJS, List as ImmutableList, Map as ImmutableMap } from "immutable";
+import classNames from "classnames";
+
+import InputFormCategoryButton from "./input-form-category-button";
+import FormField from "../form-field";
+
+import { inputForm as messages } from "../messages";
+import constants from "../constants";
+import { mayHaveAnyValue, mustHaveSomeValue } from "./validators";
+import "./input-form.scss";
+
+const Util = require("../../js/utils.js");
+
+// TEMPORARY: We define flavor hooks here for the time being.
+const MYWATER_SCHOOL_DISTRICTS = require("../../../../flavors/central-puget-sound/static/school-districts.json");
+const hooks = {
+  myWaterAddDistrict: attrs => {
+    attrs.district = MYWATER_SCHOOL_DISTRICTS[attrs["school-name"]] || "";
+
+    return attrs;
+  },
+};
+
+class InputForm extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      selectedCategory: null,
+      isCategoryMenuCollapsed: false,
+      isCategoryMenuHidden: false,
+      fields: ImmutableMap(),
+      updatingField: null,
+      isFormSubmitting: false,
+      formValidationErrors: new Set(),
+      showValidityStatus: false,
+      isMapPositioned: false,
+    };
+
+    this.customGeometryFieldname = null;
+    this.richTextFields = [];
+    this.attachments = [];
+    this.visibleCategories = [];
+
+    this.props.map.on("dragend", this.handleDragEnd.bind(this));
+  }
+
+  handleDragEnd() {
+    !this.state.isMapPositioned && this.setState({ isMapPositioned: true });
+  }
+
+  resetState() {
+    const isSingleCategory = this.visibleCategories.length === 1;
+
+    this.setState({
+      selectedCategory: isSingleCategory ? this.visibleCategories[0] : null,
+      isCategoryMenuCollapsed: false,
+      isCategoryMenuHidden: isSingleCategory,
```

Comment:
Would it make sense to make our category selection menu a separate component? 

In general, this InputForm component seems bloated and I wonder if it could be simplified by separating out the category selection logic, or at least some of the rendering details, into another component. I'm not sure, but I think the category selection menu could be implemented as a controlled, stateless component.

Reply:
Yes, I could see this being its own component.